Wildcards

avigad opened the pull request:
This fixes, extends, and closes #1579.

- adds wildcards for `simp`, `dsimp`, `dunfold`, and `delta` in interactive mode
- adds an `only` flag to suppress inclusion of default theorems in `simp`, `simph`, `simp_intros`, and `dsimp` in interactive mode.
- `mk_simp_set` is no longer private to `interactive` (it is useful for automation in `library_dev`)
- tactics that create or modify a local hypothesis return the corresponding expression: `assert`, `assertv`, `simp_at`, etc.

There are some noninteractive versions of `simp_only`, `simp_all`, etc., in `library_dev`, but let's leave them there for now, because they are mainly useful for writing automation in Lean.

jroesch committed:
feat(library/init/meta/interactive): add support for location wildcards

This allows users to now use  to apply a tactic in all hypotheses as well as the goal, we add support for all
interactive tactics using location: rewrite, simp, dsimp, unfold, and delta including each tactics many variants.

jroesch committed:
chore(tests/lean): add tests for rw and simp

avigad committed:
fix(library/init/meta/interactive): remove sorry in dunfold_occs

avigad committed:
fix(library/init/meta/simp_tactic): fix relabeling of hypotheses in simp_intro_aux

avigad committed:
fix(library/init/meta/interactive): disallow wildcard in rewrite

avigad committed:
refactor(library/init/meta/interactive): make mk_simp_set public

avigad committed:
feat(library/init/meta/interactive): add 'only' option for simp, dimph, dimp_intros, and dsimp

avigad committed:
feat(library/init/meta/tactic): make tactics that introduce a local constant return the expr

avigad committed:
feat(library/init/meta/simp_tactic): make simp_at return new hypothesis

avigad committed:
fix(library/init/meta/smt/interactive): adapt to wildcards in locations

avigad committed:
fix(tests/lean/run/simp_options): move test

avigad committed:
fix(library/init/meta/interactive): fix printing of description string for locations

Kha commented on library/init/meta/interactive.lean:
```diff
@@ -20,6 +20,18 @@ namespace interactive
+def loc.empty := loc.ns []
+
+meta instance : has_reflect loc :=
+fun l, match l with
```
nit: could also use direct function-level matching

Kha commented on library/init/meta/interactive.lean:
```diff
@@ -66,6 +81,7 @@ private meta def parser_desc_aux : expr → tactic (list format)
 | `(ident)  := return ["id"]
 | `(ident_) := return ["id"]
 | `(qexpr) := return ["expr"]
+| `(location) := return ["loc?"]
```
Did you try the default description? I can see that it may not be quite readable.

avigad replied:
The default description complained that it did not know how to display `loc.wildcard`. Can I get that with a `to_string` method? If so, I'll try it and see how it looks.

Kha replied:
Oh I see, it handles `return` but not `pure`. I pushed a fix, it should work now out of the box.

Kha commented on library/init/meta/interactive.lean:
```diff
@@ -629,12 +662,18 @@ private meta def simp_hyps (cfg : simp_config) : simp_lemmas → list name → t
+   match locat : _ → tactic unit with
+   | loc.wildcard :=
+     do ls ← local_context,
+        let loc_names := ls.map (fun l, l.local_pp_name),
```
nit: This feels like taking field notation too far :) . But I don't really mind.

leodemoura replied:
The PR also uses `fun` in a few places instead of the equivalent unicode notation that is used in the rest of the stdlib.

Kha commented on library/init/meta/interactive.lean:
```diff
@@ -37,12 +49,15 @@ meta def ident_ : parser name := ident <|> tk "_" *> return `_
 meta def with_ident_list := (tk "with" *> ident_*) <|> return []
 meta def without_ident_list := (tk "without" *> ident*) <|> return []
-meta def location := (tk "at" *> ident*) <|> return []
+meta def location := (tk "at" *> (tk "*" *> pure loc.wildcard <|> (loc.ns <$> ident*))) <|> return loc.empty
```
Reminder to myself: We should definitely introduce a `+` combinator and replace most uses of `*` with it in this file.

leodemoura replied:
@kha Yes

Kha commented on library/init/meta/interactive.lean:
```diff
@@ -655,48 +694,53 @@ It has many variants.
 
 - `simp with attr` simplifies the main goal target using the lemmas tagged with the attribute `[attr]`.
 -/
```
Doc string should be updated

Kha commented:
Now that I think about it, one may reasonably assume that `simp with,` does the same as `simp only,`, since `simp with a b` does not include the default simpset, either. It doesn't look pretty, but _would_ get us around having to reserve another token.

leodemoura commented:
> Now that I think about it, one may reasonably assume that simp with, does the same as simp only,, since simp with a b does not include the default simpset, either. It doesn't look pretty, but would get us around having to reserve another token.

@kha I agree. We already have too many keywords.  
BTW, some of them are really annoying (e.g., `take`).

avigad commented:
In the previous implementation `simp with a b` added lemmas from `a` and `b` to the default `simp _lemmas`. It sounds like you want to base it on the empty `simp_lemmas` and get rid of `only`. I will do that.

avigad committed:
refactor(library/init/meta/interactive): remove special printing of loc, style fixes

avigad committed:
fix(tests/lean/interactive/*): fix tests

avigad commented:
I made all the changes discussed above, and got rid of the 'only' flag. Now we can type
```lean
  simp [h_1, ..., h_n]
```
to simplify using the default simp set and h_1, ..., h_n, and
```lean
  simp [h_1, ..., h_n] with
```
to use *only* `h_1, ..., h_n`.

There is a problem, though. We can use
```lean
  simp [h_1, ..., h_n] with foo
```
to simplify using `h_1, ..., h_n` and the user attribute foo, or
```lean
  simp [h_1, ..., h_n] with foo bar
```
to simplify with user attributes `foo` and `bar`. But these do not include the default `simp` lemmas. Currently `simp` is not recognized as a user attribute, so we cannot write
```lean
  simp [h_1, ..., h_n] with simp foo
```
to simplify with the default simp set + foo. On the other hand, if we include `simp` lemmas by default, there is no way to exclude them.

Options:

- Make `simp` a possible attribute, so we can write
```lean
  simp [h_1, ..., h_n] with simp foo
```
  We would have to declare `simp` as an identifier at the top level or hack the parser. This seems bad.
  
- Go back to the `only` option, so `with` includes the default simp rules unless `only` is present.

- Use alterantive versions, `simp_only`, `simph_only`, `simp_intros_only`, etc., rather
  the a flag.

- Add a `use_dflt := tt` field to `simp_config`.

leodemoura commented:
> We would have to declare simp as an identifier at the top level or hack the parser. This seems bad.

Yes, it would be bad.

> Go back to the only option, so with includes the default simp rules unless only is present.

This is reasonable. I think it is the best option.

Kha commented:
> Currently simp is not recognized as a user attribute

We have `simp with foo default` for that. Undocumented, of course :) .

leodemoura commented:
> We have simp with foo default for that. Undocumented, of course :) .

@Kha Yes, you are correct.
So, are you suggesting we should not add  the `only` keyword?
@avigad wrote a nice summary of the pros and cons. To be honest, I found the syntax `simp with` a little bit awkward. I think it would be ok if we had brackets (i.e., list notation).
That is, we would write `simp with []`, `simp with [default, foo]`, etc.
What do you think?

avigad commented:
I think the solution with `default` is pretty good. We can say:
```lean
  simp [h1, ..., hn] with default, foo
  simp [h1, ..., hn] with foo
  simp [h1, ..., hn]
  simp [h1, ..., hn] with
```
and the semantics is easy to document and explain. 

With brackets:
```lean
  simp [h1, ..., hn] with [default, foo]
  simp [h1, ..., hn] with [foo]
  simp [h1, ..., hn]
  simp [h1, ..., hn] with []
```

With `only`:
```lean
  simp [h1, ..., hn] with foo
  simp only [h1, ..., hn] with foo
  simp [h1, ..., hn]
  simp only [h1, ..., hn]
```
All are o.k. by me, and I have no strong preference.

leodemoura commented:
> All are o.k. by me, and I have no strong preference.

@avigad For me, the only weird one is `simp [h1, ..., hn] with`

digama0 commented:
I vote for "simp only [h1,...,hn] with foo". It seems like a reasonable way
to recognize that "simp" is a special attribute set, and also reads fine.

On Thu, Jun 1, 2017 at 6:14 PM, Leonardo de Moura <notifications@github.com>
wrote:

> All are o.k. by me, and I have no strong preference.
>
> @avigad <https://github.com/avigad> For me, the only weird one is simp
> [h1, ..., hn] with
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/leanprover/lean/pull/1615#issuecomment-305635716>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AA1A7KHalwIm9ZPXpO0nUMh49_Iq166Jks5r_zfZgaJpZM4Npnpg>
> .
>

Kha commented:
I don't have a strong preference towards either of `only` or `with []`.

@avigad 
> We would have to declare simp as an identifier at the top level or hack the parser. This seems bad.

`simp` is not a keyword. All we would have to change is the literal in [this line](https://github.com/leanprover/lean/blob/81e97de463073c7511f01875075651259eb8c339/library/init/meta/simp_tactic.lean#L402).
I believe that would be a reasonable change independent of the surrounding issue.

avigad committed:
refactor(library/init/meta/simp_tactic): include defaults for 'simp with'

avigad committed:
feat(library/init/meta/interactive): update simp docstring

avigad committed:
fix(tests/lean/*): fix and expand tests

avigad commented:
O.k., I restored the `only` flag, added the default `[simp]` lemmas for 'with' by default, and updated the docstring. I rebased on the current version of Lean and tested, so we should be good to go.

In response to @Kha's observation, I tried getting rid of `default` in `get_user_simp_lemmas`, but this broke a test (`run/converter.lean`). It seems that at the low level, there are situations where we still need `default` (or something) to refer to the standard `[simp]` attribute. It doesn't interfere with the `with` option at all, so I left it alone.
